### ISSUE TITLE
Installing docs with `x.py install` is very slow.

### ISSUE THREAD

**Issue by vext01** [rust-lang/rust#80684]
CC @pietroalbini

On today's git (0cd459f) `x.py install` seems to get stuck if you are building docs.

In `config.toml`:
```
[install]
prefix = "build/ykrustc-stage2-latest"
sysconfdir = "etc"
```

`x.py install` will get stuck at:
```
    Finished release [optimized] target(s) in 8.38s
Dist rust-docs-1.51.0-dev-x86_64-unknown-linux-gnu
        finished in 17.222 seconds
Install docs stage2 (Some(TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None }))
install: creating uninstall script at /home/vext01/source/rust/build/tmp/empty_dir/build/ykrustc-stage2-latest/lib/rustlib/uninstall.sh
install: installing component 'rust-docs'
```

The underlying command is: `sh /home/vext01/source/rust/build/tmp/tarball/rust-docs/x86_64-unknown-linux-gnu/rust-docs-1.51.0-dev-x86_64-unknown-linux-gnu/install.sh --prefix=/home/vext01/source/rust/build/ykrustc-stage2-latest --sysconfdir=build/ykrustc-stage2-latest/etc --datadir=build/ykrustc-stage2-latest/share --docdir=build/ykrustc-stage2-latest/share/doc/rust --bindir=build/ykrustc-stage2-latest/bin --libdir=build/ykrustc-stage2-latest/lib --mandir=build/ykrustc-stage2-latest/share/man --disable-ldconfig`

Adding this to the config seems to work around the hang:
```
[build]
docs = false
```

Thanks!

**Comment by pietroalbini** [comment]
How long did you let the command run before killing it? I remember installing the docs taking a *long* time last time I did it.

**Comment by vext01** [comment]
I left it for half an hour or so IIRC. I'll run it for longer and see if it goes away.

How long do you think it should take to install on a modern SSD?

**Comment by pietroalbini** [comment]
I don't remember exactly how long it takes, but it doesn't depend on the speed of the disk. `install.sh` is just really inefficient. You can see if it's actually hanging by checking the contents of `lib/rustlib/install.log` this command while it's installing:

```
tail -f build/ykrustc-stage2-latest/lib/rustlib/install.log
```

**Comment by vext01** [comment]
The log file is empty, but I can see new files appearing on the disk (using `find` and `diff`) so progress must be occurring. Slowly.

**Comment by vext01** [comment]
It's been about 40 minutes now. It seems to be adding files to `build/tmp/empty_dir/` is that right?

**Comment by vext01** [comment]
> Build completed successfully in 0:57:44

So it's just very slow.

**Comment by jyn514** [comment]
I took a look into this. The main issue is that the generated `install.sh` is very inefficient; it takes about 5 minutes to copy 2500 files (there are more than 2500, but I got bored and cancelled it). @pietroalbini @Mark-Simulacrum what do you think about rewriting this inner loop in Rust so it's not launching so many external processes?

https://github.com/rust-lang/rust-installer/blob/5254dbfd25d5284728ab624dca1969d61427a0db/install-template.sh#L548

**Comment by daira** [comment]
This issue is compounded on macOS, where it seems to be even more inefficient to launch lots of processes from a shell. Sorry I don't have timings, but it takes tens of minutes at least for this step; long enough that I would expect most users who are trying to build our software from source in macOS (it's Zcash, which installs a pinned version of Rust in order to build) are likely to think it's hanging and give up.

**Comment by Mark-Simulacrum** [comment]
I think the problem may largely be that the manifest.in generated for rust-docs component (at least) is a file: list instead of using the directory: command; directory commands should be much more efficient. This change may be a recent(ish) regression; I suspect that fixing that is likely the right solution here.

**Comment by daira** [comment]
We solved our problem for Zcash by just not building rust-docs (passing `--without=rust-docs` to `install.sh`). Obviously that doesn't solve the underlying issue.

**Comment by mbargull** [comment]
FYI, https://github.com/rust-lang/rust-installer/pull/106 addresses this (not the underlying cause of the inefficiency but the majority of its extent).

**Comment by andersk** [comment]
I bisected the underlying issue to c768ce138427b1844c1f6594daba9c0e33928032 “bootstrap: convert rust-docs to use Tarball” (#79788).

Before this commit, rust-docs installs in half a second, and `rust-docs/manifest.in` is 1 line:

```
dir:share/doc/rust/html
```

After this commit, rust-docs installs in some twenty minutes, and `rust-docs/manifest.in` is 18742 lines:

```
file:share/doc/rust/html/rust.css
file:share/doc/rust/html/version_info.html
file:share/doc/rust/html/complement-design-faq.html
file:share/doc/rust/html/complement-lang-faq.html
file:share/doc/rust/html/complement-project-faq.html
file:share/doc/rust/html/guide-crates.html
file:share/doc/rust/html/guide-ffi.html
file:share/doc/rust/html/guide-macros.html
file:share/doc/rust/html/guide-strings.html
file:share/doc/rust/html/guide-tasks.html
…
file:share/doc/rust/html/rustc/fonts/source-code-pro-v11-all-charsets-500.woff2
file:share/doc/rust/html/rustc/searchindex.json
file:share/doc/rust/html/rustc/searchindex.js
file:share/doc/rust/html/rustc/searcher.js
file:share/doc/rust/html/rustc/mark.min.js
file:share/doc/rust/html/rustc/elasticlunr.min.js
file:share/doc/rust/html/rustc/images/image1.png
file:share/doc/rust/html/rustc/images/image2.png
file:share/doc/rust/html/rustc/images/image3.png
file:share/doc/rust/html/robots.txt
```

**Comment by andersk** [comment]
Proposed fix in #84289. (We should also do the optimization in rust-lang/rust-installer#106 of course.)
